### ISSUE TITLE
Transient write on vector of boxes

### ISSUE THREAD

**Issue by raygray** [arximboldi/immer#33]
I'm trying to use vector of boxes and everything woks fine until transient write is used. 
```cpp
using Element = immer::box<std::string>;
auto vect = immer::vector<Element>;

// this one works fine
for (auto i = 0; i < 100; ++i) {
    vect = vect.push_back(Element("x"));
}

// this one doesn't compile
auto t = vect.transient();

for (auto i = 0; i < 100; ++i) {
    t.push_back(Element("x"));
}

vect = t.persistent();
```
Any ideas?

**Comment by arximboldi** [comment]
@raygray, can you post the compiler output? And compiler version/os. Thanks!

**Comment by raygray** [comment]
This code clarifies the problem a bit:
```cpp
struct Val {
};

using Element = immer::box<Val>;

immer::vector<Element> vect;

auto t = vect.transient();

for (auto i = 0; i < 100; ++i) {
    t.push_back(Element());
}

// compilation error here
vect = t.persistent();
```
Both gcc and clang throw similar error:
```
error : excess elements in struct initializer
       holder(Args&&... args) : value{std::forward<Args>(args)...} {}
```
By 'unwinding' all the error messages, it seems like `persistent_type{ impl_ };` (in the `vector_transient::persistent()` function) gets 'confused' when `immer::vector<immer::box>` is used...

**Comment by arximboldi** [comment]
Thanks for the clarification.  I will look into that.  One question: why
are you boxing the elements?  In general it is more efficient to store
the elements unboxed, unless you want to pass around lots of copies of
Val and is big and non-POD.

Cheers!

  JP

raygray <notifications@github.com> writes:

> This code clarifies the problem a bit:
> ```cpp
> struct Val {
> };
>
> using Element = immer::box<Val>;
>
> immer::vector<Element> vect;
>
> auto t = vect.transient();
>
> for (auto i = 0; i < 100; ++i) {
>     t.push_back(Element());
> }
>
> // compilation error here
> vect = t.persistent();
> ```
> Both gcc and clang throw similar error:
> ```
> error : excess elements in struct initializer
>        holder(Args&&... args) : value{std::forward<Args>(args)...} {}
> ```
> By 'unwinding' all the error messages, it seems like `persistent_type{ impl_ };` (in the `vector_transient::persistent()` function) gets 'confused' when `immer::vector<immer::box>` is used...
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/arximboldi/immer/issues/33#issuecomment-372891664

-- 
∿∿∿  https://sinusoid.al
∿∿∿  https://sinusoid.es

**Comment by arximboldi** [comment]
Addressed by #41, can you @raygray that this solves your problems once it lands master?
